File: src/language/Swift.py
import re


def extract_libraries(files):
    """Extracts a list of imports that were used in the files

    Parameters
    ----------
    files : []string
        Full paths to files that need to be analysed

    Returns
    -------
    dict
        imports that were used in the provided files, mapped against the language
    """

    res = []
    # regex to find imports like import Stuff
    regex_import = re.compile(r"import ((?!func|var|let|typealias|protocol|enum|class|struct)\s?[a-zA-Z0-9.]+)", re.IGNORECASE)

    # regex to find imports like import kind module.symbol
    # here kind can be func, var, let, typealias, protocol, enum, class, struct
    # this regex will return a list of (kind, module.symbol)
    regex_declarations = re.compile(r"import (func|var|let|typealias|protocol|enum|class|struct) ([a-zA-Z0-9.]+)", re.IGNORECASE)

    for f in files:
        with open(file=f, mode='r', errors='ignore') as fr:

            contents = ' '.join(fr.readlines())
            matches = regex_import.findall(contents)
            matches.extend([d[1] for d in regex_declarations.findall(contents)])

        if matches:
            res.extend(matches)

    # remove duplicates
    res = list(set(res))
    return {"Swift": res}

File: librarydetection/languages/Swift.go
package languages

import (
	"regexp"

	"github.com/codersrank-org/repo_info_extractor/librarydetection"
)

// NewSwiftAnalyzer constructor
func NewSwiftAnalyzer() librarydetection.Analyzer {
	return &swiftAnalyzer{}
}

type swiftAnalyzer struct{}

func (a *swiftAnalyzer) ExtractLibraries(contents string) ([]string, error) {
	// regex to find imports like import Stuff
	regexImport, err := regexp.Compile(`import ([a-zA-Z\.]*)[\n|\r\n]`)
	if err != nil {
		return nil, err
	}
	// regex to find imports like import kind module.symbol
	// here kind can be func, var, let, typealias, protocol, enum, class, struct
	// this regex will return a list of (kind, module.symbol)
	regexDeclarations, err := regexp.Compile(`import [func|var|let|typealias|protocol|enum|class|struct]+ ([a-zA-Z|\.]*)`)
	if err != nil {
		return nil, err
	}

	return executeRegexes(contents, []*regexp.Regexp{regexImport, regexDeclarations}), nil
}
